runtime/lua/vim/treesitter/highlighter.lua

```diff
@@ -278,86 +283,119 @@ end
-      local capture, node, metadata, match = state.iter(line)
-
-      local range = { root_end_row + 1, 0, root_end_row + 1, 0 }
-      if node then
-        range = vim.treesitter.get_range(node, buf, metadata and metadata[capture])
+    while cmp_lt(next_row, next_col, range_er, range_ec) do
+      local capture, node, metadata, match = state.iter(range_er, range_ec)
+      if not node then
+        next_row = math.huge
+        next_col = math.huge
+        break
       end
-      local start_row, start_col, end_row, end_col = Range.unpack4(range)
-
-      if capture then
-        local hl = state.highlighter_query:get_hl_from_capture(capture)
 
-        local capture_name = state.highlighter_query:query().captures[capture]
-
-        local spell, spell_pri_offset = get_spell(capture_name)
-
-        -- The "priority" attribute can be set at the pattern level or on a particular capture
-        local priority = (
-          tonumber(metadata.priority or metadata[capture] and metadata[capture].priority)
-          or vim.hl.priorities.treesitter
-        ) + spell_pri_offset
-
-        -- The "conceal" attribute can be set at the pattern level or on a particular capture
-        local conceal = metadata.conceal or metadata[capture] and metadata[capture].conceal
+      local range = vim.treesitter.get_range(node, buf, metadata and metadata[capture])
+      local start_row, start_col, end_row, end_col = Range.unpack4(range)
 
-        local url = get_url(match, buf, capture, metadata)
+      if cmp_lt(next_row, next_col, start_row, start_col) then
+        next_row = start_row
+        next_col = start_col
+      end
 
-        if hl and end_row >= line and (not is_spell_nav or spell ~= nil) then
-          api.nvim_buf_set_extmark(buf, ns, start_row, start_col, {
-            end_line = end_row,
-            end_col = end_col,
-            hl_group = hl,
-            ephemeral = true,
-            priority = priority,
-            conceal = conceal,
-            spell = spell,
-            url = url,
-          })
-        end
+      if not capture then
+        break
       end
 
-      if start_row > line then
-        state.next_row = start_row
+      local hl = state.highlighter_query:get_hl_from_capture(capture)
+      local capture_name = state.highlighter_query:query().captures[capture]
+      local spell, spell_pri_offset = get_spell(capture_name)
+
+      -- The "priority" attribute can be set at the pattern level or on a particular capture
+      local priority = (
+        tonumber(metadata.priority or metadata[capture] and metadata[capture].priority)
+        or vim.hl.priorities.treesitter
+      ) + spell_pri_offset
+
+      -- The "conceal" attribute can be set at the pattern level or on a particular capture
+      local conceal = metadata.conceal or metadata[capture] and metadata[capture].conceal
+
+      local url = get_url(match, buf, capture, metadata)
+
+      if not is_spell_nav or spell ~= nil then
```

Comment:
Since `hl` is always a number, the first check was always true. 

Proper check would be `hl ~= 0 or conceal or url`. Otherwise the tests fail since the marks aren't added.

I removed the check since without highlighting conceal or url, the decor ranges wouldn't be added anyway.